2018/002-update-channels.md

```diff
@@ -0,0 +1,184 @@
+update is ready to be installed. They should run on each worker node to be more independent
+and decoupled from velum. If velum would be the initiator of the update check, it might lead
+to inconsistent and long execution runs in larger clusters.
+
+* "check for updates" service on each worker
+  - reports availability of maintenance updates to velum via grains
+  - once per day e.g.
+  - decoupled from velum, runs independently, just reports
+* "new version" checker service on each worker
+  - check for new versions (migration targets)
+    - query RMT/SMT/SCC API if there is a new compatible version available for this CaaSP version (SUSEConnect library)
+    - report back to velum via grains
+  - get the link to the release notes
+  - once per week or longer timeframe
+  - decoupled from velum, runs independently, just reports
+
+#### User Interface
+
+The user interface of velum needs to be adapted minimally.
+
+* [indirect] display of update content through [a link to the] release notes
+* display of update history via velum
+* button to trigger the maintenance updates (what we currently have)
+* button to switch to the next channel in velum
+* separate UI for updates and upgrades (new migration target)
+
+#### API (optional)
+
+Furthermore there could be an API available through velum to enable easy scripting
+
+* endpoints:
+  - GET:  list current channel maintenance updates (non disruptive)
+  - POST: trigger current channel maintenance updates
+  - GET:  list next channel (new version) updates
+      - A: disruptive update in the same caasp product/version (e.g. k8s minor version bump, container runtime)
+      - B: migration target once overlap support is EOL (next caasp product/version)
+  - POST: switch to the next channel
+  - GET:  list update history
+* we would need a way of CLI authentication to get to this API
+
+#### General considerations
+
+##### Being outdated before a migration
+
+We need to check if the installed packages are good enough to do the migration or not.
+If it's good enough, we can either continue with the migration directly or apply normal
+updates and prolong the migration.
+The user should not be enforced to run all maintenance updates first, because the migration
+would contain/overwrite the patches anyways.
+
+##### Out of sync nodes
+
+There needs to be a way to get a single node up-to-date with maintenance updates and
+new migration targets, especially when a single node is out of sync with updates, due to
+having been powered off for a longer time, or if the customer added a new node from
+an old ISO image to the cluster.
+
+##### Systems Registered against RMT/SMT
+
+If the system is registered against SMT or RMT, we can access the information, that there
+is a new version, but that this version is not completely mirrored to SMT/RMT.
+So we need an additional information: a new version is available, but the admin needs to
+mirror this first on SMT/RMT.
```

Comment:
I think this makes sense, but again I would not do that as part of the 1st iteration

Reply:
I'm afraid we need to do that for the first iteration (and it should be trivial to display this message), else update will fail later with spurious errors.
Using SMT/RMT is very common, even all of our public cloud images depends on this.

Reply:
According to what is written into the RFC (a couple of lines above):

> * "new version" checker service on each worker
>   - check for new versions (migration targets)
>     - query RMT/SMT/SCC API if there is a new compatible version available for this CaaSP version (SUSEConnect library)
>     - report back to velum via grains
>   - get the link to the release notes
>   - once per week or longer timeframe
>   - decoupled from velum, runs independently, just reports

I think we don't even need to care about that. If the node is registered against SMT/RMT the "new version"
 check service will not query SCC, but will talk with SMT/RMT. Now it's up to SMT/RMT to report the availability of a new version **only** when all the needed channels are mirrored.